shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs

```diff
@@ -0,0 +1,326 @@
+        keyRegKesPeriod
+        ocert
+
+instance Arbitrary Mock.BHeader where
+  arbitrary = do
+    res <- arbitrary :: Gen Mock.Block
+    return $ case res of
+      Block header _ -> header
+      _ -> error "SerializationProperties::BHeader - failed to deconstruct header from block"
+
+instance Crypto c => Arbitrary (HashHeader c) where
+  arbitrary = HashHeader <$> genHash (Proxy @c)
+
+instance Arbitrary Mock.Tx where
+  arbitrary = do
+    (_ledgerState, _steps, _txfee, tx, _lv) <- hedgehog genStateTx
+    return tx
+
+instance Crypto c => Arbitrary (TxId c) where
+  arbitrary = TxId <$> genHash (Proxy @c)
+
+instance Crypto c => Arbitrary (TxIn c) where
+  arbitrary =
+    TxIn
+      <$> (TxId <$> genHash (Proxy @c))
+      <*> arbitrary
+
+instance Arbitrary Mock.TxOut where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary Nonce where
+  arbitrary =
+    oneof
+      [ return NeutralNonce,
+        mkNonce . fromIntegral <$> choose (1, 123 :: Int)
+      ]
+
+instance Arbitrary UnitInterval where
+  arbitrary = fromJust . mkUnitInterval . (% 100) <$> choose (1, 99)
+
+instance Crypto c => Arbitrary (KeyHash a c) where
+  arbitrary = KeyHash <$> genHash (Proxy @c)
+
+instance Arbitrary Natural where
+  arbitrary = fromInteger <$> choose (0, 1000)
+
+instance Arbitrary (STS.PredicateFailure Mock.PPUP) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.UTXO) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.UTXOW) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.POOL) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.DELPL) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.DELEG) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.DELEGS) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.LEDGER) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary (STS.PredicateFailure Mock.LEDGERS) where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance Arbitrary Coin where
+  -- Cannot be negative even though it is an 'Integer'
+  arbitrary = Coin <$> choose (0, 1000)
+
+instance Arbitrary SlotNo where
+  -- Cannot be negative even though it is an 'Integer'
+  arbitrary = SlotNo <$> choose (1, 100000)
+
+instance Arbitrary EpochNo where
+  -- Cannot be negative even though it is an 'Integer'
+  arbitrary = EpochNo <$> choose (1, 100000)
+
+instance Arbitrary Mock.Addr where
+  arbitrary =
+    oneof
+      [ Addr <$> arbitrary <*> arbitrary <*> arbitrary
+      -- TODO generate Byron addresses too
+      -- SL.AddrBootstrap
```

Comment:
Don't forget about this TODO that was copied.

Reply:
@JaredCorduan see here for when you're done with adding Byron addrs